packages/backend/server/src/core/config/service.ts

```diff
@@ -1,17 +1,110 @@
-import { Injectable } from '@nestjs/common';
-import { PrismaClient } from '@prisma/client';
+import { Injectable, OnApplicationBootstrap } from '@nestjs/common';
+import { once, set } from 'lodash-es';
+
+import { ConfigLoader, EventBus, OnEvent } from '../../base';
+import { APP_CONFIG_DESCRIPTORS } from '../../base/config/register';
+import { Models } from '../../models';
+import { ServerFeature } from './types';
+
+declare global {
+  interface Events {
+    'config.init': {
+      config: DeepReadonly<AppConfig>;
+    };
+    'config.changed': {
+      updates: DeepPartial<AppConfig>;
+    };
+    'config.changed.broadcast': {
+      updates: DeepPartial<AppConfig>;
+    };
+  }
+}
 
 @Injectable()
-export class ServerService {
+export class ServerService implements OnApplicationBootstrap {
   private _initialized: boolean | null = null;
-  constructor(private readonly db: PrismaClient) {}
+  readonly #features = new Set<ServerFeature>();
+
+  constructor(
+    private readonly models: Models,
+    private readonly configLoader: ConfigLoader,
+    private readonly event: EventBus
+  ) {}
+
+  async onApplicationBootstrap() {
+    await this.setup();
+  }
+
+  get features() {
+    return Array.from(this.#features);
+  }
 
   async initialized() {
     if (!this._initialized) {
-      const userCount = await this.db.user.count();
+      const userCount = await this.models.user.count();
       this._initialized = userCount > 0;
     }
 
     return this._initialized;
   }
+
+  enableFeature(feature: ServerFeature) {
+    this.#features.add(feature);
+  }
+
+  disableFeature(feature: ServerFeature) {
+    this.#features.delete(feature);
+  }
+
+  get config() {
+    return this.configLoader.config;
+  }
+
+  getConfigDescriptors = once(() => {
+    return Object.fromEntries(
+      Object.entries(APP_CONFIG_DESCRIPTORS).map(([module, descriptors]) => [
+        module,
+        Object.entries(descriptors).map(([key, descriptor]) => ({
+          key,
+          ...descriptor,
+        })),
+      ])
+    );
+  });
+
+  async updateConfig(
+    user: string,
+    updates: Array<{ key: string; value: any }>
+  ) {
+    const promises = await this.models.appConfig.save(user, updates);
+
+    const overrides: DeepPartial<AppConfig> = {};
+    // only take successfully saved configs
+    promises.forEach(promise => {
+      if (promise.status === 'fulfilled') {
+        set(updates, promise.value.id, promise.value.value);
+      }
+    });
+    this.configLoader.override(overrides);
+    this.event.emit('config.changed', { updates: overrides });
+    this.event.broadcast('config.changed.broadcast', { updates: overrides });
```

Comment:
If a node restarts or crashes during the `broadcast`, or for some special reason does not listen to this `broadcast` event, is there a possibility of configuration inconsistency?

Once this situation occurs, we need to wait until the next process restart to pull the configuration from the database to achieve eventual consistency.

At this stage, this low-probability event should still be acceptable; in fact, this is a function of the configuration service.

Reply:
add a cron to fetch latest config every 5 mins?

Reply:
I just happened to think of this potential issue, so let's ignore it for now. 
Since the change frequency is very low, the probability of it occurring is even lower. Let's wait and see a better solution if a problem really arises later.